Ensure snippets are displayed in a package’s settings view.

savetheclocktower opened the pull request:
Add an “Enable” checkbox inside the snippet view to add/remove the current package from the snippets blacklist.

### Description of the Change

`PackageSnippetsView` is rehabilitated and made to show a package’s snippets (should a package have any snippets to display). The issues described in #1076 have been fixed.

An “Enable” checkbox has been added at the top of the section — identical to the one at the top of a Keybindings panel. Just as unchecking the Keybindings panel’s Enable checkbox immediately disables that package’s key bindings, unchecking the Snippet panel’s Enable checkbox immediately disables all snippets provided by that package. No reloading of the window is necessary.

The snippets themselves are listed in the table with columns for prefix, name, scope, and body. Since the body of a snippet is arbitrary text of arbitrary length, we don’t show it in the table; we show it in a tooltip that is shown when the <kbd>View</kbd> button is clicked. The <kbd>Copy</kbd> button writes a string to the clipboard that can be pasted into a user’s `snippets.cson` (or `snippets.json`) for further customization.

### Alternate Designs

I mooted this change in #1076 before contributing this PR because I wasn’t yet certain how snippet bodies ought to be displayed in the UI.

### Benefits

First of all, this fixes the bug (described in #1076) that prevented a package's settings page from listing snippets altogether.

This is also the UI side of atom/snippets#277, which allows package-provided snippets to be disabled on a per-package basis. This is one of the oldest and most-requested enhancements for the snippets package.

### Possible Drawbacks

Can't think of any. This fixes a regression. We were already paying the cost of retrieving the snippets for display in `PackageSnippetsView`; we were merely failing to display them.

### Applicable Issues

* #1076 has discussion about the bugfix and the UI for the snippets table.
* atom/snippets#55 is the original feature request, and atom/snippets#277 is **a prerequisite change** for this PR. Ideally these two PRs should be landed simultaneously, but under no circumstances should this one land before the other, or else this UI will be hooked up to nothing at all.

savetheclocktower commented:
The CI failures are happening because a couple of the specs rely on behavior added in atom/snippets#277.

daviwil commented on lib/package-snippets-view.js:
```diff
@@ -76,11 +109,25 @@ export default class PackageSnippetsView {
-  addSnippets () {
+  getSnippetsModule () {
+    const snippetsPackage = atom.packages.getLoadedPackage('snippets')
+    return snippetsPackage ? snippetsPackage.mainModule : null
```
You can find the service object definition [here](https://github.com/atom/snippets/blob/e9fc6eece2ad7142c070ce9e502e1a045a030e7e/lib/snippets.coffee#L420), just exposing the `getUserSnippetsPath` should be good enough!

savetheclocktower replied:
Changing the reference to `mainModule` in this added section is easy enough. There is preexisting code in `getSnippets` that deals with the module directly instead of the exposed service object — will that need to get addressed before we can land this, or can I disregard it for now?

Ben3eeE replied:
@daviwil Do you think we should open an issue for this?

https://github.com/atom/settings-view/blob/57b4f01cb2ece1815a9eb0494227daf5b735f11c/lib/package-snippets-view.js#L100

daviwil replied:
Interesting, I didn't realize there's pre-existing code for this.  I just took a look at how we consume the snippet service and it appears we don't even pass along the object directly but instead pass along its `getSnippets` method with a different object:

https://github.com/atom/settings-view/blob/master/lib/main.js#L86-L90

It looks like this code definitely needs to be overhauled at some point in the future so I'd suggest leaving the current usage of `snippetsPackage.mainModule` the way it is for now.

If it's not too much extra work to get the user settings service method piped through, it'd be ideal if we could at least do that part well without the rest of it being fixed.  If it ends up adding more headache than it's worth, I'd be OK with just calling directly into the `mainModule` to get the config path.

savetheclocktower replied:
> If it's not too much extra work to get the user settings service method piped through, it'd be ideal if we could at least do that part well without the rest of it being fixed. If it ends up adding more headache than it's worth, I'd be OK with just calling directly into the mainModule to get the config path.

Nah, that part was easy enough. The rest I’d be happy to tackle in a separate issue.

daviwil replied:
Excellent, thanks a lot!

Ben3eeE commented:
@simurai Do you have any feedback on the design here? I saw you commented in https://github.com/atom/settings-view/issues/1076 and said it looked good.

Ben3eeE commented:
@savetheclocktower I hope to get this merged so it goes out in beta in Atom 1.33 🙂 but I can't make any promises about that. Please version guard it for Atom 1.33 🙇

daviwil commented:
We'll be creating the release branches for 1.32.0 and 1.33.0-beta0 on Friday.  If this PR gets in before then, checking for a minor version `>= 33` should be sufficient, otherwise you'll have to check for 1.34.0.

savetheclocktower commented:
OK, the last commit addresses all feedback. @Ben3eeE, I fixed the issue with newlines when copying to the clipboard, and I changed the specs to cover that scenario. But since those tests are now version-guarded they won't run unless you temporarily change the spec file just to verify it.

(Also updated atom/snippets#277 to expose `getUserSnippetsPath` on the provider.)

Ben3eeE commented:
Huh I'm getting this error when trying to test out the changes. I don't know why I definitely pulled and linked both the settings-view and snippets pr. I'll check more tomorrow.

```
Uncaught TypeError: this.snippetsProvider.getUserSnippetsPath is not a function
    at PackageSnippetsView.writeSnippetToClipboard (/I:/atom_master/settings-view/lib/package-snippets-view.js:212)
```

I noticed another thing. The [bundled snippets in the snippets package](https://github.com/atom/snippets/blob/master/lib/snippets.cson) do not show up in the settings for the snippets package so those can't be disabled or copied. I am not sure if they have ever shown up in the settings. This is not something that we have to address now and can open an issue for.

daviwil commented on spec/installed-package-view-spec.coffee:
```diff
@@ -40,6 +40,10 @@ describe "InstalledPackageView", ->
 
+    # Relies on behavior not present in the snippets package before 1.33.
+    # TODO: These tests should always run once 1.33 is released.
+    shouldRunScopeTest = parseFloat(atom.getVersion()) >= 1.33
```
This is slightly horrifying but I'm surprised to find that it works pretty well for this purpose even with versions like `1.33.0-beta3` :)  Since we're using the pattern in other repos I think it's fine to use it here too.

savetheclocktower replied:
> Huh I'm getting this error when trying to test out the changes. I don't know why I definitely pulled and linked both the settings-view and snippets pr. I'll check more tomorrow.

I bet that's a legit error. It occurred to me that I didn’t actually test out the functionality after that change — I only ran the specs. But the specs use a mock snippets provider, so I haven't tested the integration at all. I'll look at that shortly.

Ben3eeE replied:
Haha, yeah. We get many commenting about this. And also the fact that 1.5 > 1.33. But we still use it :)

savetheclocktower commented:
> Huh I'm getting this error when trying to test out the changes. I don't know why I definitely pulled and linked both the settings-view and snippets pr. I'll check more tomorrow.

I bet that's a legit error. It occurred to me that I didn’t actually test out the functionality after that change — I only ran the specs. But the specs use a mock snippets provider, so I haven't tested the integration at all. I'll look at that shortly.

savetheclocktower commented:
> I noticed another thing. The bundled snippets in the snippets package do not show up in the settings for the snippets package so those can't be disabled or copied. I am not sure if they have ever shown up in the settings. This is not something that we have to address now and can open an issue for.

I _think_ this happens for any package that is currently `apm link`'d in dev mode. The package's path is the symlinked version (`~/.atom/dev/packages/foo`), but the snippets it should be recognizing as belonging to package `foo` are all reporting the actual, resolved path of `foo`'s snippet's file.

savetheclocktower commented:
Anyway: fixed the error @Ben3eeE reported (settings-view wraps the snippetsProvider for some reason) but CI is still being ornery. I’m not sure why it fails on AppVeyor. The previous iteration passed on stable but failed on beta, and now it's failing both.

savetheclocktower commented:
Oh, the test in question is flaky (see #973). I feel slightly better about this.

Ben3eeE commented:
It's green after rebuilding twice :+1: Seems like this test is very flaky. I'll pull and test all the changes later today to see that everything works before merging.

Ben3eeE commented on lib/package-snippets-view.js:
```diff
@@ -110,19 +150,86 @@ export default class PackageSnippetsView {
+    'body': '${body}'
+`
+    } else {
+      body = body.replace(/"/g, `\\"`).replace(/\n/g, '\\n')
```
Oops. I completely forgot that tabs should be replaced with `\t` when writing my review comment about newlines. The article snippet copies the newlines correctly but the tab character is still there.

daviwil commented:
Kicked off the AppVeyor build again to see if it goes green.  Is there anything else that needs to be done for this PR?

Ben3eeE commented:
@savetheclocktower Thank you for contributing 🙇 This is going to be a great addition 💯
